tests/unittest.py

```diff
@@ -124,6 +129,74 @@ def new(*args: P.args, **kwargs: P.kwargs) -> R:
     return _around
 
 
+@overload
+def deepcopy_config(config: RootConfig, root: Literal[True]) -> RootConfig:
+    ...
+
+
+@overload
+def deepcopy_config(config: Config, root: Literal[False]) -> Config:
+    ...
+
+
+def deepcopy_config(
+    config: Union[Config, RootConfig], root: bool
+) -> Union[RootConfig, Config]:
+    new_config: Union[Config, RootConfig]
```

Comment:
While I don't fully understand the need for this function, I think what you probably want to do here is 

```python
C = TypeVar("C", Config, RootConfig)
# or if that doesn't work, C = TypeVar("C", bound=Union[Config, RootConfig])
def deepcopy_config(config: C) -> C: ...
```

and replace `if root` with `if isinstance(config, RootConfig)`. That should help avoid the casts.

Reply:
Looks like mypy can't figure it out:

```
_TConfig = TypeVar("_TConfig", Config, RootConfig)


def deepcopy_config(config: _TConfig) -> _TConfig:
    new_config: _TConfig

    if isinstance(config, RootConfig):
        reveal_type(config.__class__)
        new_config = config.__class__(config.config_files)
    else:
        new_config = config.__class__(config.root)
```

prints:

```
$ poetry run mypy tests/unittest.py
tests/unittest.py:138: note: Revealed type is "Type[tests.unittest.<subclass of "Config" and "RootConfig">]"
tests/unittest.py:138: note: Revealed type is "Type[synapse.config._base.RootConfig]"
tests/unittest.py:139: error: Argument 1 to <subclass of "Config" and "RootConfig"> has incompatible type "List[str]"; expected "Optional[RootConfig]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

I'll tell it to ignore the issue